test/tools/yulInterpreter/Interpreter.cpp

```diff
@@ -368,3 +418,80 @@ void ExpressionEvaluator::incrementStep()
 		BOOST_THROW_EXCEPTION(ExpressionNestingLimitReached());
 	}
 }
+
+void ExpressionEvaluator::runExternalCall(evmasm::Instruction _instruction)
+{
+	u256 memOutOffset = 0;
+	u256 memOutSize = 0;
+	u256 callvalue = 0;
+
+	// Setup memOut* values
+	if (
+		_instruction == evmasm::Instruction::CALL ||
+		_instruction == evmasm::Instruction::CALLCODE
+	)
+	{
+		memOutOffset = values()[5];
+		memOutSize = values()[6];
+		callvalue = values()[2];
+	}
+	else if (
+		_instruction == evmasm::Instruction::DELEGATECALL ||
+		_instruction == evmasm::Instruction::STATICCALL
+	)
+	{
+		memOutOffset = values()[4];
+		memOutSize = values()[5];
+	}
+	else
+		yulAssert(false);
+
+	// Don't execute external call if it isn't our own address
+	if (values()[1] != util::h160::Arith(m_state.address))
+		return;
+
+	Scope tmpScope;
+	InterpreterState tmpState;
+	tmpState.calldata = m_state.calldata;
+	tmpState.callvalue = callvalue;
+
+	// Create new interpreter for the called contract
+	unique_ptr<Interpreter> newInterpreter = makeInterpreterNew(tmpState, tmpScope);
+
+	Scope* abstractRootScope = &m_scope;
+	Scope* fileScope = nullptr;
+	Block const* ast = nullptr;
+
+	// Find file scope
+	while (abstractRootScope->parent)
+	{
+		fileScope = abstractRootScope;
+		abstractRootScope = abstractRootScope->parent;
+	}
+
+	// Get AST for file scope
+	for (auto&& [block, scope]: abstractRootScope->subScopes)
+		if (scope.get() == fileScope)
+		{
+			ast = block;
```

Comment:
If I understand correctly, `ast` will be set to the block that is containing the bloch the will be called, right? But somehow I don't understand how you exactly find that block that is called. Maybe it would be helpful - at least for me - to add an additional comment how you exactly find the block that need to get called.

Reply:
It reads like that you're iterating through all current scopes and if the scope is the file scope, you set the `ast`. Somehow I don't see that the found block, is really just the block of the function that need to be called. For me it looks like that it could also be another block. The problem is that I cannot understand the code correctly I guess.

Reply:
There are two steps here:

First we find the scope that equals our whole file. This scope can be found directly under the "virtual" root scope which is just there to collect all the other scopes.
We can't access it directly, however as we only have our current scope, so we search upwards until we find the very end (when parent == nullptr), then our scope must be the root scope and the one before must be our file scope.

Once we have our file scope, we want the actual already parsed AST which is only available in a mapping from block -> scope, so we have to manually iterate through that mapping to find our AST.

In yul, the highest-level AST for files is of type block, that's why I called it that way here